### DISCUSSION TITLE
AGH and Browser AdBlocker

### DISCUSSION THREAD

**Discussion by gthb96** [AdguardTeam/AdGuardHome#2884] [Q&A] [answered]
Hello,
I have a question that bugs me for some time.
Let's assume I have AGH running and use an AdBlocker in my Browser.

Do cosmetic filters in the browser still work?
If yes how can they work if the ad is not able to be downloaded?
Like how does the Browser AdBlocker know what to change in the layout of the website if it doens't know how the ad is fomatted?

**Comment by ameshkov** [comment, marked as answer]
> Do cosmetic filters in the browser still work?

Sure, they do.

> If yes how can they work if the ad is not able to be downloaded?

Not every ad can be blocked on the DNS level.

1. Google search ads - they're not loaded from a particular URL are simply embedded in the seach results
2. Facebook ads - same thing, embedded in the feed
3. Also, there're examples of ads that are loaded from the same domain as the content, DNS-level blocker cannot do anything with it.

> Like how does the Browser AdBlocker know what to change in the layout of the website if it doens't know how the ad is fomatted?

This is not an automated process. Browsers ad blockers use filter lists composed by filters maintainers, and they add rules that target websites' layout.

If you want to know more about traditional filter lists, I recommend checking out this doc: https://kb.adguard.com/en/general/how-to-create-your-own-ad-filters

Careful, it's huge:)

**Reply by gthb96** [reply]
Thanks so much for your answer!
But I think my question is not completely answered yet.

Think about a particular ad that is being blocked on the DNS level.
Can the browser AdBlocker in this case still do the cosmetic part?

**Reply by gthb96** [reply]
I think I got it now.
Even if an ad is blocked on DNS level the cosmetic filter works.
I think that even if no DNS filter is running the browser blocker itself prevents loading of the ad from the server.

And because of the page sources the cosmetic filter knows where the ad would be and does it's magic.

Is my assumption approximately correct?

**Reply by ameshkov** [reply]
Pretty much, yes
